invoices/invoiceregistry.go

```diff
@@ -439,113 +439,25 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
-	updateSubscribers := false
-
-	updateInvoice := func(inv *channeldb.Invoice) (
-		*channeldb.InvoiceUpdateDesc, error) {
-
-		// Don't update the invoice when this is a replayed htlc.
-		htlc, ok := inv.Htlcs[circuitKey]
-		if ok {
-			switch htlc.State {
-			case channeldb.HtlcStateCanceled:
-				debugLog("replayed htlc to canceled invoice")
-
-			case channeldb.HtlcStateAccepted:
-				debugLog("replayed htlc to accepted invoice")
-
-			case channeldb.HtlcStateSettled:
-				debugLog("replayed htlc to settled invoice")
-
-			default:
-				return nil, errors.New("unexpected htlc state")
-			}
-
-			return nil, errNoUpdate
-		}
-
-		// If the invoice is already canceled, there is no further
-		// checking to do.
-		if inv.State == channeldb.ContractCanceled {
-			debugLog("invoice already canceled")
-			return nil, errNoUpdate
-		}
-
-		// If an invoice amount is specified, check that enough
-		// is paid. Also check this for duplicate payments if
-		// the invoice is already settled or accepted.
-		if inv.Terms.Value > 0 && amtPaid < inv.Terms.Value {
-			debugLog("amount too low")
-			return nil, errNoUpdate
-		}
-
-		// The invoice is still open. Check the expiry.
-		if expiry < uint32(currentHeight+i.finalCltvRejectDelta) {
-			debugLog("expiry too soon")
-			return nil, errNoUpdate
-		}
-
-		if expiry < uint32(currentHeight+inv.Terms.FinalCltvDelta) {
-			debugLog("expiry too soon")
-			return nil, errNoUpdate
-		}
-
-		// Record HTLC in the invoice database.
-		newHtlcs := map[channeldb.CircuitKey]*channeldb.HtlcAcceptDesc{
-			circuitKey: {
-				Amt:          amtPaid,
-				Expiry:       expiry,
-				AcceptHeight: currentHeight,
-			},
-		}
-
-		update := channeldb.InvoiceUpdateDesc{
-			Htlcs: newHtlcs,
-		}
-
-		// Don't update invoice state if we are accepting a duplicate
-		// payment. We do accept or settle the HTLC.
-		switch inv.State {
-		case channeldb.ContractAccepted:
-			debugLog("accepting duplicate payment to accepted invoice")
-			update.State = channeldb.ContractAccepted
-			return &update, nil
-
-		case channeldb.ContractSettled:
-			debugLog("accepting duplicate payment to settled invoice")
-			update.State = channeldb.ContractSettled
-			return &update, nil
-		}
-
-		// Check to see if we can settle or this is an hold invoice and
-		// we need to wait for the preimage.
-		holdInvoice := inv.Terms.PaymentPreimage == channeldb.UnknownPreimage
-		if holdInvoice {
-			debugLog("accepted")
-			update.State = channeldb.ContractAccepted
-		} else {
-			debugLog("settled")
-			update.Preimage = inv.Terms.PaymentPreimage
-			update.State = channeldb.ContractSettled
-		}
-
-		updateSubscribers = true
-
-		return &update, nil
-	}
-
 	// We'll attempt to settle an invoice matching this rHash on disk (if
 	// one exists). The callback will set the resolution action that is
 	// returned to the link or contract resolver.
-	invoice, err := i.cdb.UpdateInvoice(rHash, updateInvoice)
+	updater := invoiceUpdater{
```

Comment:
personal style preference: I think encapsulating these in a struct hides a little bit what is going on with the callback. What do you think about something like

```go
invoice, err := i.cdb.UpdateInvoice(
    rHash, InvoiceUpdateCallback  {
     return update( ... params ....)
})
```

Reply:
There are quite a few parameters (another one will be added for mpp later) and also two return values atm. It could be rewritten, but I favor this approach.

Reply:
Reconsidering this again. Hold off re-review

Reply:
I came around to this approach. Updated.